Merge polynomial interpolator branch

joezuntz opened the pull request:
I think we are now ready for a PR here to address issue #4 at least in part. There are a few things missing, but to get something in place for early tests of the machinery I hope this should be suitable.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+from numpy.polynomial.laguerre import lagval2d
+from numpy.polynomial.hermite import hermval2d
+
+POLYNOMIAL_TYPES = {
```
Would you mind using lower case for this (both the dict and the keys)?  Uppercase is pretty uncommon in python.  Usually reserved for constant numerical values (although `math.pi` and `numpy.pi`).  It seems like lower case would be the more typical style for this.

joezuntz replied:
PEP8 has upper case for module-level constants, which is what I was following here, but happy to change to whatever you prefer.

rmjarvis replied:
This feels more like a "Global Variable" rather than a "Constant", in which case PEP8 recommends lower_case.  After all, there would be nothing wrong with a user adding more items to this dict if they had another kind of polynomial they function they wanted to use.  (Not that I'm a huge PEP8 stickler by any stretch -- there are a number of places where I prefer to depart from them.)

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+    surface to each parameter passed in independently.
+
+    :param orders:  List/array of integers, one for each parameter 
+                    to be interpolated.
```
Missing documentation of `poly_type`.  Should mention the `polynomial_types` dict for the user to see what the valid options are.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+        self.set_function(poly_type)
+        self.coeffs = None
+
+    def set_orders(self, orders):
```
Probably this is not a function we want the user to use directly.  Should give it a leading underscore to mark it as a "private" member function.  (Of course, nothing is actually private in Python, but this is a standard convention.)

rmjarvis replied:
Likewise for some of the other methods below.  `set_function`, `generate_indices`, `pack_coefficients`, etc.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+        TODO: At the moment this function is expecting a numpy array of
+        shape (2,nstar) for the positions. We might want to use a galsim
+        position object instead since I think some code elsewhere in Piff
+        is expecting this.
```
We should probably make our own Position class to encapsulate the "position" used for interpolation.  e.g. it may include the color or other such information.  And it should know which values constitute a regular (u,v) part of the position (and so need to use the max i+j order) vs. using having its own separate polynomial order.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+        #Call the appropriate function to generate the polynomial model.
+        #By default this is numpy.polyval. Note that despite appearances
+        #this is not a method call - function is a normal python attribute
+        #that happens to be a function.
```
A minor style point.  I find it hard to read comments that don't have a space after the `#`.  Would you mind adding a space?  (Here and elsewhere in this file.)

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+        :param positions:  A list of positions ((u,v) in this case) of stars.
+        :param parameter:  A numpy array of the measured values of a parameter
+                           for each star
+                           
```
`:param parameter_index:`?

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+        #to convert these to numpy arrays and transpose
+        #them to the order we need.
+        parameters = numpy.array(vectors).T
+        positions = numpy.array(pos).T
```
It would be nice to log some information here.  Something like
```
if logger:
    logger.info("Fitting %d parameter vectors using polynomial type %s",len(vectors),self.poly_type)
```
Probably some extra detail at the `debug` level too would be helpful.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,361 @@
+            # Black boxes curve fitter from scipy!
+            # We may want to look into the tolerance and other parameters
+            # of this function.
+            p,covmat=scipy.optimize.curve_fit(model, positions, parameter, p0)
```
I'm a little surprised that we need a non-linear solver here.  Isn't this a straightforward matrix inversion?  Solving Ax = b -> x = (AtA)^-1 (Atb).  Is it just because the numpy polynomial functions don't give you any way to go from C -> A?  That the only interface is through that functional form `f(u,v,C)`?

cpadavis replied:
I think in general it ends up being faster / more stable. This was the same thing matt did in osprey.

We should also be able to include the weight map in the interpolation fit, which should just be as simple as:

    # weights are the same shape as parameters
    sigma = np.zeros_like(weights)
    msk = weights > 0
    sigma = np.sqrt(1.0/weights[msk])
    parameter = parameter[msk]
    positions=positions[msk]
    p,covmat=scipy.optimize.curve_fit(model, positions, parameter, p0, sigma=sigma, absolute_sigma=True)

rmjarvis commented on tests/test_poly.py:
```diff
@@ -0,0 +1,287 @@
+    #Check that the shape is 4*4 in the unpacked (because we
+    #want space for all the terms), and that we can unpack and
+    #repack successfully.
+    assert (packed==packed_test).all()
```
numpy.testing has some nicer ways to write tests involving numpy arrays.
```
numpy.testing.assert_array_equal(packed_test, packed)
```
Then for things that don't need to be exact, there are `assert_almost_equal` and similar things that tend to be useful.

rmjarvis commented on tests/test_poly.py:
```diff
@@ -0,0 +1,287 @@
+    #we can pack and then unpack
+    packed_test_2 = interp.pack_coefficients(0,unpacked_test)
+    unpacked_test_2 = interp.unpack_coefficients(0,packed_test_2)
+    assert (unpacked_test_2 == unpacked_test).all()
```
I'm confused about this test.  How could anything be non-zero here?  I'm not sure what this is supposed to be testing.

joezuntz replied:
This is testing that the polynomial coefficients with total order i+j greater than the maximum total order 3 are zero. It's testing unpacked[i,j]==0, which is not identically true since it's unpacked_test that was created just above with zeros_like, rather than unpacked; unpacked was created on line 38 by unpacked a random vector.

joezuntz replied:
I will clarify with a comment.

rmjarvis commented on tests/test_poly.py:
```diff
@@ -0,0 +1,287 @@
+        assert numpy.allclose(v, mean)
+
+
+def sub_poly_linear(type1):
```
I think nosetests will only run functions that start with (or maybe include?) `test`.  So should change the name of this function.

rmjarvis replied:
Oh, never mind.  This is run from the below function.  Sorry for the confusion.

rmjarvis commented on tests/test_poly.py:
```diff
@@ -0,0 +1,287 @@
+
+def test_poly_quadratic():
+    #This is basically the same as linear but with
+    #quadratic variation
```
Any reason not to do this the same way as linear, with all the types in `PolynomialsTypes`?

joezuntz replied:
Good point! There was a reason originally when I was testing actual coeff values but not any more.

rmjarvis commented on tests/test_poly.py:
```diff
@@ -0,0 +1,287 @@
+        poly_load_save_sub(poly_type,poly_type)
+
+def test_poly_load_err():
+    from nose.tools import assert_raises
```
Is this a relic?  Did you used to have this raise an exception, but now it overwrites the existing?

joezuntz replied:
Yes - deleting it now.

rmjarvis commented on tests/test_poly.py:
```diff
@@ -0,0 +1,287 @@
+    for poly_type1 in PolynomialsTypes[:]:
+        for poly_type2 in PolynomialsTypes[:]:
+            if poly_type1!=poly_type2:
+                poly_load_save_sub(poly_type1,poly_type2)
```
Should put a bit at the bottom that lets you run this from the command line:

if __name__ == '__main__':
    test_poly_indexing()
    test_poly_mean()
    test_poly_linear()
    test_poly_quadratic()
    ...
```

joezuntz commented:
Okay, think I have addressed all of these comments now, apart from the curve_fit thing as it seemed like that was the way to go, and the class to encapsulate positions as that needs more work elsewhere in the code first.  Let me know if there's anything else to do.

joezuntz commented:
Also, many thanks! This is a really useful exercise, as always. I wish all my code had someone reading it in this level of depth.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,389 @@
+        n=self.orders[parameter_index]+1
+        C = numpy.zeros((n, n))
+        for k,(i,j) in enumerate(self.indices[parameter_index]):
+                C[i,j] = coeffs[k]
```
Minor comment here, but this seems to have gotten extra indent spaces for some reason.

rmjarvis commented on piff/polynomial_interp.py:
```diff
@@ -0,0 +1,389 @@
+        # We should have the same number of parameters as number of polynomial 
+        # orders with which we were created here.
+        nparam = len(parameters)
+        assert nparam==self.nparam, """Must create Polynomial interpolator 
```
I think assert is the wrong thing here.  Asserts should be things that are always true if the code doesn't have bugs.  Not a way to check for user errors.  I'd rather this be
```
if nparam != self.nparam:
    raise ValueError("Must create Polynomial interpolator with ...")
```

rmjarvis commented:
Thanks, Joe.  Just a couple minor things now. 

So, on this branch too, would anyone else like to do a code review?

joezuntz commented:
Okay, cool - I've fixed those and the tests still pass. I think we'll want to revisit this whole thing later with more sophisticated models for what coefficients we have for different parameters but hopefully it's useful for now.

cpadavis commented:
looked good to me
